ext/MatrixAlgebraKitMooncakeExt/MatrixAlgebraKitMooncakeExt.jl

```diff
@@ -0,0 +1,225 @@
+module MatrixAlgebraKitMooncakeExt
+
+using Mooncake
+using Mooncake: DefaultCtx, CoDual, Dual, NoRData, rrule!!, frule!!, arrayify, @is_primitive
+using MatrixAlgebraKit
+using MatrixAlgebraKit: inv_safe, diagview, copy_input
+using MatrixAlgebraKit: qr_pullback!, lq_pullback!
+using MatrixAlgebraKit: qr_null_pullback!, lq_null_pullback!
+using MatrixAlgebraKit: eig_pullback!, eigh_pullback!
+using MatrixAlgebraKit: left_polar_pullback!, right_polar_pullback!
+using LinearAlgebra
+
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(copy_input), Any, Any}
+function Mooncake.rrule!!(::CoDual{typeof(copy_input)}, f_df::CoDual, A_dA::CoDual)
+    Ac = copy_input(Mooncake.primal(f_df), Mooncake.primal(A_dA))
+    dAc = Mooncake.zero_tangent(Ac)
+    function copy_input_pb(::Mooncake.NoRData)
+        Mooncake.increment!!(Mooncake.tangent(A_dA), dAc)
+        return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+    end
+    return CoDual(Ac, dAc), copy_input_pb
+end
+
+# two-argument factorizations like LQ, QR, EIG
+for (f, pb, adj) in (
+        (qr_full!, qr_pullback!, :dqr_adjoint),
+        (qr_compact!, qr_pullback!, :dqr_adjoint),
+        (lq_full!, lq_pullback!, :dlq_adjoint),
+        (lq_compact!, lq_pullback!, :dlq_adjoint),
+        (eig_full!, eig_pullback!, :deig_adjoint),
+        (eigh_full!, eigh_pullback!, :deigh_adjoint),
+        (left_polar!, left_polar_pullback!, :dleft_polar_adjoint),
+        (right_polar!, right_polar_pullback!, :dright_polar_adjoint),
+    )
+
+    @eval begin
+        @is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof($f), Any, Tuple{<:Any, <:Any}, MatrixAlgebraKit.AbstractAlgorithm}
+        function Mooncake.rrule!!(::CoDual{typeof($f)}, A_dA::CoDual, args_dargs::CoDual, alg_dalg::CoDual{<:MatrixAlgebraKit.AbstractAlgorithm}; kwargs...)
```

Comment:
```suggestion
        function Mooncake.rrule!!(::CoDual{typeof($f)}, A_dA::CoDual, args_dargs::CoDual, alg_dalg::CoDual{<:MatrixAlgebraKit.AbstractAlgorithm})
```

I don't think we are allowing any additional keyword arguments on top of what is already in the `AbstractAlgorithm`, so this probably does not need to be here

Reply:
That's definitely a good question. What I really meant is that we are not allowing any extra kwargs to the primal calls, which I think has as a consequence that there shouldn't be kwargs here to pass on? I agree that we will at some point have to figure out how to control these settings, but I'm open to suggestions there because I definitely haven't figured that out.